File: [78] Subsets/78.subsets.cpp
/*
 * @lc app=leetcode id=78 lang=cpp
 *
 * [78] Subsets
 */

// @lc code=start
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> output;

    vector<vector<int>> subsets(vector<int>& nums) {
        // using backtracking algorithm
        int nums_len = nums.size();
        for (int c = 0; c <= nums_len; c++) {
            vector<int> current;
            backtrack(nums, current, 0, c);
        }
        return output;
    }

    void backtrack(vector<int>& nums, vector<int>& current, int start, int capacity) {
        // capacity is the capacity of current subset to be filled
        if (current.size() == capacity) {
            output.emplace_back(current);
            return;
        } else {
            for (int i = start; i < nums.size(); i++) {
                current.emplace_back(nums[i]);
                backtrack(nums, current, i+1, capacity);
                current.pop_back();
            }
        }
    }
};
// @lc code=end

File: [78] Subsets/78.subsets.java
import java.util.ArrayList;

/*
 * @lc app=leetcode id=78 lang=java
 *
 * [78] Subsets
 */

// @lc code=start
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        res.add(new ArrayList<>());
        cascading(nums, res, 0);
        return res;
    }

    public void cascading(int[] nums, List<List<Integer>> res, int current_idx) {
        if (current_idx == nums.length) return;
        int res_len = res.size();
        for (int i = 0; i < res_len; i++) {
            List<Integer> current = new ArrayList<>(res.get(i));
            current.add(nums[current_idx]);
            res.add(current);
        }
        cascading(nums, res, current_idx+1);
    }
}
// @lc code=end

